examples/ex24p.cpp

```diff
@@ -0,0 +1,334 @@
+   HYPRE_Int size = fespace->GlobalTrueVSize();
+   HYPRE_Int H1size = H1fespace->GlobalTrueVSize();
+   if (myid == 0)
+   {
+      cout << "Number of Nedelec finite element unknowns: " << size << endl;
+      cout << "Number of H1 finite element unknowns: " << H1size << endl;
+   }
+
+   // 7. Define the solution vector x as a parallel finite element grid function
+   //    corresponding to fespace. Initialize x by projecting the exact
+   //    solution. Note that only values from the boundary edges will be used
+   //    when eliminating the non-homogeneous boundary condition to modify the
+   //    r.h.s. vector b.
+   ParGridFunction x(fespace);
+   FunctionCoefficient p_coef(p_exact);
+   ParGridFunction p(H1fespace);
+   p.ProjectCoefficient(p_coef);
+
+   VectorFunctionCoefficient gradp_coef(sdim, gradp_exact);
+
+   // 8. Set up the parallel bilinear form corresponding to the EM diffusion
+   //    operator curl muinv curl + sigma I, by adding the curl-curl and the
+   //    mass domain integrators.
+   Coefficient *muinv = new ConstantCoefficient(1.0);
+   Coefficient *sigma = new ConstantCoefficient(1.0);
+   ParBilinearForm *a = new ParBilinearForm(fespace);
+   ParMixedBilinearForm *a_NDH1 = new ParMixedBilinearForm(H1fespace, fespace);
+   if (pa)
+   {
+      a->SetAssemblyLevel(AssemblyLevel::PARTIAL);
+      a_NDH1->SetAssemblyLevel(AssemblyLevel::PARTIAL);
+   }
+
+   a->AddDomainIntegrator(new VectorFEMassIntegrator(*sigma));
+
+   a_NDH1->AddDomainIntegrator(new MixedVectorGradientIntegrator(*muinv));
+
+   // 9. Assemble the parallel bilinear form and the corresponding linear
+   //    system, applying any necessary transformations such as: parallel
+   //    assembly, eliminating boundary conditions, applying conforming
+   //    constraints for non-conforming AMR, static condensation, etc.
+   if (static_cond) { a->EnableStaticCondensation(); }
+
+   a->Assemble();
+   if (!pa) { a->Finalize(); }
+
+   a_NDH1->Assemble();
+   if (!pa) { a_NDH1->Finalize(); }
+
+   Vector B(fespace->GetTrueVSize());
+   Vector X(fespace->GetTrueVSize());
+
+   if (pa)
+   {
+      a_NDH1->Mult(p, x);
+      x.GetTrueDofs(B);
+   }
+   else
+   {
+      HypreParMatrix *NDH1 = a_NDH1->ParallelAssemble();
+
+      Vector P(H1fespace->GetTrueVSize());
+      p.GetTrueDofs(P);
+
+      NDH1->Mult(P,B);
+
+      delete NDH1;
+   }
+
+   // 10. Define and apply a parallel PCG solver for AX=B with no
+   //     preconditioner, in the full assembly case.
+   //     With partial assembly, use Jacobi preconditioner.
+
+   if (pa)
+   {
+      ParGridFunction diag_pa(fespace);
+      Vector tdiag_pa(fespace->GetTrueVSize());
+      a->AssembleDiagonal(tdiag_pa);
+      diag_pa.SetFromTrueDofs(tdiag_pa);
+
+      Array<int> ess_tdof_list;
+      OperatorJacobiSmoother Jacobi(diag_pa, ess_tdof_list, 1.0);
+
+      CGSolver cg(MPI_COMM_WORLD);
+      cg.SetRelTol(1e-12);
+      cg.SetMaxIter(1000);
+      cg.SetPrintLevel(1);
+      cg.SetOperator(*a);
```

Comment:
This is not quite right -- the action of `ParBilinearForm::Mult` is purely local, so the system that is solved here is decoupled across processors. That is one reason to use `FormLinearSystem` and then use the returned `OperatorPtr`.

The `Jacobi` smoother should be setup on the true dofs as well.

Reply:
Thanks for catching this. It was getting the correct solution, but not correctly. This is fixed now.